### DISCUSSION TITLE
Composite Primary key

### DISCUSSION THREAD

**Discussion by AaronRiboy** [serenity-is/Serenity#6617] [Help] [answered]
[image contains sensitive data please mask them]

Hello there, i have 2 composite primary key in the table. Does anyone know how to solve it this. Thank you

**Comment by edwardch** [comment, marked as answer]
I had a composite (two) key in an old project and I remember adding both fields to my Row.cs, then a third field that was an expression that put them together, and set that as my IdProperty.

Something like this 

Expression("CAST(T0.DeviceId AS VARCHAR(12)) + CAST(T0.Timestamp AS VARCHAR(20))")

This was about four years ago and there is probably a better way of doing it.

**Reply by AaronRiboy** [reply]
The IdProperty u added as a new field?

**Reply by edwardch** [reply]
Yes so three fields:
- Prop1 (Part of composite key)
- Prop2 (Part of composite key)
- Prop3 (only add this to Row.cs, use an Expression, and set as IdProperty)

Another option that has been suggested before, in #279, is to add another column to the table called Id, as this doesn't have to be a part of the primary key. All depends what you want to achieve with them...

**Reply by AaronRiboy** [reply]
<img width="516" alt="image" src="https://user-images.githubusercontent.com/29535650/208582458-914d2784-8969-45b6-ac4e-1caa4ad4adc8.png">
I did try your method bro. I got error when i run my program. MasterId and DealerID is a primary key.Added new value as Property and set it as IdProperty 

<img width="391" alt="image" src="https://user-images.githubusercontent.com/29535650/208582750-67a04f96-979b-42b7-851b-926afbdf0191.png">

**Reply by edwardch** [reply]
What version of Serenity is this? Also I see you have two fields marked as `PrimaryKey`. I just checked and I don't have any properties marked as PrimaryKey in my project, so maybe try remove both and see what you get?

**Reply by AaronRiboy** [reply]
<img width="576" alt="image" src="https://user-images.githubusercontent.com/29535650/208604881-b3c433b4-fa44-4acf-b886-9ad10b7fda03.png">
**Its 6.0.8 v(serenity.net.web)**
Previously ,IdProperty was on master dealer,once added back to the master dealer no more error is showing but the composite key error still remain 
<img width="877" alt="image" src="https://user-images.githubusercontent.com/29535650/208605975-3da8f26c-e7d8-4324-9cb5-62b59408e9ba.png">

**Reply by edwardch** [reply]
Can you remove the IdProperty from the MasterDealer property. 

Serenity's default grid requires the Id to be unique, and in your case you have the same master dealer for two (or more) different dealers. Which is normal. Your PropertyId is the combination of master dealer and dealer, and that will be unique.

**Reply by AaronRiboy** [reply]
Bro, its work..Thank you so much
